### DISCUSSION TITLE
Implementing a custom driver

### DISCUSSION THREAD

**Discussion by farnoy** [tokio-rs/tokio#4180] [Q&A] [answered]
Hi,

I'm interested in building a custom driver to have my Futures suspend in a non-blocking way where I'm able to wake the ones that are ready from a separate thread outside of the runtime. I've looked at how the `time` driver works, and if I understand correctly, any time that the Runtime cannot make further progress (because all Futures are Pending), it lets the drivers "park" to do something driver-specific and have it wake up specific futures so that they're polled again. Is that the gist of it?

I'd like to do something like that, but it's application-specific and I don't want to fork/modify the tokio library.

I think the workaround is to have my centralized logic run a spin loop on a separate thread and continuously check the status of outstanding Wakers to see which ones I can activate. I won't be explicitly parked by tokio, so I have to anticipate new Wakers being submitted to my pseudo-driver, which is why I'd need an inefficient spin loop to notice these additions.

The specific call I want to wrap is [vkWaitSemaphores](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/chap7.html#vkWaitSemaphores), which is a blocking call that accepts multiple objects (so multiple suspended futures can be batched into one call) with a timeout. It can be called with "wait for all" semantics like `join!` or "wait any" which corresponds to `select!`/`Promise.race()`.

**Comment by Darksonn** [comment, marked as answer]
Your idea of having a separate thread is a good one. Basically, you would write a custom impl of the `Future` trait that asks your dedicated thread to add the semaphore to its list of semaphores, also informing the thread which waker it should wake when the semaphore gets a permit. There is an example of waking a future from another thread [here][1].

As for noticing new additions to the set of semaphores, why not make a semaphore for that? Every time you've added a new waker, you add a permit on a special semaphore, which wakes up your thread. The thread can then check for new semaphores and go back to sleep.

[1]: https://tokio.rs/tokio/tutorial/async

**Reply by farnoy** [reply]
Thanks for the tip. I didn't realize that I can use the inputs to that blocking call to set up an unparking gadget, but it's working very well. I'm able to set the maximum timeout in that call and only unpark the thread when new futures are polled that require this wait (or when previous ones are ready, of course).

I'm struggling with lifetimes and ownership, but if I manage to clean it up, I'll post an example for others.

**Reply by Darksonn** [reply]
Regarding ownership, my main tip is that you want the shared data between the futures and threads to be in an `Arc<Mutex<...>>` that each future and the thread has a clone of.
